### ISSUE TITLE
[CM P0] Kotlin binding (extension of Java binding)

### ISSUE THREAD

**Issue by mandoway** [ElektraInitiative/libelektra#4294]
# Elektra Kotlin Binding proposal

>### This is only a proposal, feel free to suggest any changes !
>#### Also **please** comment any usecases you would like to see 'kotlinized' :)
>Proposal from Philipp Leeb (@Gratla) and Markus Bointner (@mandoway)

Kotlin can be seen as a 'successor' to Java in a way, since it also compiles to JVM bytecode.
It has some interesting features not available in Java which we want to take advantage of.
Therefore, we propose an **extension** of the Elektra Java binding which can utilise Kotlin's functionalities.

We aim to implement the following tasks:

- [ ] DSL for creating Keys and KeySets (@Gratla)
- [x] Key getter/setter with type inference (@mandoway)
    - [x] Custom classes via serialization
    - [x] Collections via serialization
- [x] Iterating over key parts (e.g. "/key/to/value" -> ["key", "to", "value"]) (@mandoway)
- [ ] Easy error and warning creation (@Gratla)

Excluded:
- New kotlin binding with Kotlin Native (we extend the JNA binding)
- Re-implementations of features in JNA binding (we can always call Java code from Kotlin)

---
We plan to create a subproject in the JNA binding folder to simplifiy the build process.   
The binding will also be built with gradle and will use the JNA library as a base.

---
To get a better understanding what we want to acheive, the following sections provide usage examples of our envisioned extensions: 

### Functional paradigm

We can abstract the try-with-resources access for kdb using lambdas in Kotlin:

General kdb usage:
Java:
```java
try (KDB kdb = KDB.open()) {
    // code to manipulate keys
    kdb.get(..);
    kdb.set(..);
} catch (KDB.KDBException e) {
    e.printStackTrace();
}
```

Kotlin:
```kotlin
withKdb { 
    get(..)
    set(..)
}

```

PluginLoader in Java:
```java
PluginLoader pluginLoader = new PluginLoader();
Plugin rangePlugin = pluginLoader.loadElektraPlugin("range");
Key rangeKey = Key.create("user:/tests/myError", "30");
rangeKey.setMeta("check/range", "1-20");
final KeySet ks = KeySet.create(10, KeySet.KS_END);
ks.append(rangeKey);
rangePlugin.kdbSet(ks, parentKey);
```

In Kotlin:
```kotlin
withPlugin("range") { 
    val rangeKey = keyOf("user:/tests/myError" to "30").apply {
        setMeta("check/range", "1-20")
        // or alternatively
        meta = "check/range" to "1-20"
    }
    // or 
    val rangeKey = keyOf(
        value = "user:/tests/myError" to "30",
        meta = "check/range" to "1-20"
    )
    
    val ks = keySetOf(allocationHint = 10, rangeKey)
    
    kdbSet(ks, parentKey)
}
```

Shifting the imperative style of Java to a more Kotlin-idiomatic and functional style.
We omitted the error handling in this example, but the function `withKdb` would also throw the KDBException.

### Type inference and assertion
Kotlin also improves the type system in comparison to Java.
We envision to implement the following:

```java
key.getString();
key.getBoolean();
key.setString();
key.setBoolean();
```

```kotlin
key.get<String>()
key.get<Boolean>()

// Automatic type inference
val foo: Boolean = key.get()
key.set("bar")
key.set(true)

data class MyProperties(val foo: String, val bar: Boolean)

// Existing keys in kdb:
// user:/parent/foo
// user:/parent/bar
val props: MyProperties = keySetOf("user:/parent").get()
val props = keySetOf("user:/parent").as<MyProperties>()
val props = fromKdb<MyProperties>("user:/parent")
val props = kdb.decode<MyProperties>("user:/parent")
val props = kdb.build<MyProperties>("user:/parent")
```

### Functional KeySet traversal
```java
var keyNamespace = Key.create("user:/");           // select a namespace from which all keys should be fetched
try (KDB kdb = KDB.open()) {
    var keySet = kdb.get(keyNamespace);            // fetch all keys into a new key set
    for (var currentKey : keySet) {                // traverse the set
        System.out.println(String.format("%s: %s",
                currentKey.getName(),              // fetch the key's name
                currentKey.getString()));          // fetch the key's value
    }
} catch (KDB.KDBException e) {
    e.printStackTrace();
}
```

```kotlin
// New ways to create keys and keySets
// Error handling omitted
keySetOf(keyOf("user:/")).forEach { key ->
    println("${key.name}: ${key.get<String>}")
}

// Mapping a whole keyset to another value
// We could also add a shortcut for keySetOf
keySetOf("user:/").map { key ->
    key.name.uppercase()
}

keySetOf("user:/").filter {
    it.isBinary
}

// Or even shorter
keySetOf("user:/").filter<Boolean>().forEach{ key ->
    println(key)
}

keySet.forEach { key ->
    // Key: /foo/bar = bloo
    key.parts.forEach { part -> 
        // Prints: foo bar
        print(part)
    }
}

keySet.values.forEach { value ->
    print(value)
}
```

**Comment by markus2330** [comment]
The issue sounds very promising! :rocket: 

> Type inference and assertion

Does this also work with arrays and custom types (classes)?

> Functional KeySet traversal

Iteration over key names would be also interesting!

Do you want to implement T2 also in Kotlin? Then probably further use cases will come up, like a convenient way to create errors and warnings.

**Comment by kodebach** [comment]
> Does this also work with arrays and custom types (classes)?

Using [kotlinx.serialization](https://github.com/Kotlin/kotlinx.serialization) might also be an idea. There is standard functionality to transform from/to [JSON](https://kotlin.github.io/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/-json/decode-from-json-element.html), so it should in theory be possible to do something similar for a `KeySet`.

> Iteration over key names would be also interesting!

To clarify, "iterating over key names" means iterating over the parts of a key name, as is done already in e.g. C++.

**Comment by Gratla** [comment]
We like the idea of using the type inference also for collections.
As of now, the implementation with serializeable objects should be possible and we want to include this topic into this issue.  
Also the iteration over key names is a fine addition.

**Comment by mandoway** [comment]
@markus2330 

We updated the issue according to the feedback:
- Added responsible team member for tasks
- Added excluded features
- Added implementation suggestion for CMakeLists

**Comment by kodebach** [comment]
> We plan to create a separate directory in the bindings folder and create a new CMakeLists file inspired by the one in JNA.

I think it may be better to just create a Gradle sub-project in the existing `src/bindings/jna`, i.e. just create a folder `src/bindings/jna/kotlin` or similar and add it to `settings.gradle`. Otherwise the dependency between Java binding and Kotlin binding might be unecessarily complicated to implement.

The code base will soon be restructured anyway (in the summer, once CM is done). So keeping to the simplest possible solution regarding folder structure and CMake setup is probably best.

**Comment by markus2330** [comment]
@tucek are you fine with having the Kotlin binding within the jna binding?

@kodebach I don't know about a restructure of the jna binding in the summer. Do you plan to do it?

**Comment by kodebach** [comment]
>  I don't know about a restructure of the jna binding in the summer. Do you plan to do it?

I was talking about the `library_directory_structure` decision in #4243. With that structure we'd have a top-level (in `src`) Gradle setup. So then I'd move the Kotlin stuff to `src/bindings/kotlin`, because then it is easier to add the dependency on the underlying Java binding (since it's still the same Gradle project).

I referred to the summer, because I assumed we won't touch the CMake/Gradle/etc. files during the semester. (At the time I wrote the comment, it also wasn't clear that there wouldn't be more changes to the directory structure)

> we might exclude the Kotlin binding from build task and provide a separate buildKotlin task.

That is probably the best solution for now. Alternatively, we just exclude it and require manually building the sub-project for Kotlin.
